modAionImpl/src/org/aion/zero/impl/sync/msg/RequestTrieData.java

```diff
@@ -0,0 +1,134 @@
+package org.aion.zero.impl.sync.msg;
+
+import java.math.BigInteger;
+import java.util.Objects;
+import org.aion.p2p.Ctrl;
+import org.aion.p2p.Msg;
+import org.aion.p2p.Ver;
+import org.aion.rlp.RLP;
+import org.aion.rlp.RLPList;
+import org.aion.zero.impl.sync.Act;
+import org.aion.zero.impl.sync.DatabaseType;
+
+/**
+ * Request message for a trie node from a specific blockchain database.
+ *
+ * @author Alexandra Roatis
+ */
+public final class RequestTrieData extends Msg {
+    private final DatabaseType dbType;
+    private final byte[] nodeKey;
+    private final int limit;
+
+    /**
+     * Constructor for trie node requests with specified limit.
+     *
+     * @param nodeKey the key of the requested trie node
+     * @param dbType the blockchain database in which the key should be found
+     * @param limit the maximum number of key-value pairs to be retrieved by the search inside the
+     *     trie for referenced nodes
+     * @throws NullPointerException if either of the given parameters is {@code null}
+     * @throws IllegalArgumentException if the given limit is negative
+     */
+    public RequestTrieData(final byte[] nodeKey, final DatabaseType dbType, final int limit) {
+        super(Ver.V1, Ctrl.SYNC, Act.REQUEST_TRIE_DATA);
+
+        // ensure inputs are not null
+        Objects.requireNonNull(nodeKey);
+        Objects.requireNonNull(dbType);
+
+        // ensure limit is positive
+        if (limit < 0) {
+            throw new IllegalArgumentException(
+                    "The RequestTrieData object cannot be built with a negative limit.");
+        }
+
+        this.nodeKey = nodeKey;
+        this.dbType = dbType;
+        this.limit = limit;
+    }
+
+    /**
+     * Decodes a message into a trie node request.
+     *
+     * @param message a {@code byte} array representing a request for a trie node.
+     * @return the decoded trie node request if valid or {@code null} when the decoding encounters
+     *     invalid input
+     * @implNote Ensures that the components are not {@code null}.
+     */
+    public static RequestTrieData decode(final byte[] message) {
+        if (message == null || message.length == 0) {
+            return null;
+        } else {
+            RLPList list = (RLPList) RLP.decode2(message).get(0);
+            if (list.size() != 3) {
```

Comment:
Is this an unexpected case?  If so, should it log an error/warning?

Reply:
The handlers introduced in #817 and #818 will log errors in case decoding the messages was not possible. I don't think "the why" is meaningful. The essential here is to discard corrupt messages.